Updated Sming to SDK 2.1

slaff opened the pull request:
* The good: The code and samples compile for SDK 1.5, 2.0 and 2.1+
* The bad: Unfortunately in samples that use SDK 2.1+ we can see roughly 5K less available free memory.
* The ugly: SDK 2.1+ comes with wrong definitions for core functions, like `ets_delay_us`, that gives the impression that SDK 2.1+ is not really top quality.

anakod commented:
@slaff I have many issues with building.  Before I have used old stable Sming version with few my own pathces (will publish it after).
To check your commint, I have installed last UDK version, download and replace SDK from https://github.com/espressif/ESP8266_NONOS_SDK switch to your branch, and after that I have these compilation errors:

```
20:01:25 **** Build of configuration Sming for project SmingFramework ****
c:/Espressif/ESP8266_SDK/include/osapi.h:32:6: error: conflicting types for 'ets_bzero'
 void ets_bzero(void *s, size_t n);

...

include/arch/cc.h:49:15: error: conflicting types for 'ets_strlen'
 extern size_t ets_strlen(const char *s);

...

include/arch/cc.h:52:13: error: conflicting types for 'ets_timer_arm_new'
 extern void ets_timer_arm_new(ETSTimer *ptimer, uint32_t milliseconds, bool repeat_flag, int isMstimer);

...

make[2]: *** [lwip/core/def.o] Error 1
make[2]: Leaving directory `/C/DEV/ESP8266/Sming/Sming/third-party/esp-open-lwip'
make[1]: *** [compiler/lib/liblwip_open.a] Error 2
make[1]: Leaving directory `/C/DEV/ESP8266/Sming/Sming'
make: *** [reload] Error 2

20:01:30 Build Finished (took 4s.980ms)
```

I tryed to remove openlwip folder and rebuild all. Sming compiled successfuly.

Now I'm trying to build Basic_Serial example:

```
make[1]: Entering directory `/C/DEV/ESP8266/Sming/Sming'
In file included from third-party/pwm/pwm.c:48:0:
third-party/pwm/pwm.c: In function 'pwm_init':
c:/Espressif/ESP8266_SDK/include/ets_sys.h:78:41: error: passing argument 2 of 'ets_isr_attach' from incompatible pointer type [-Werror=incompatible-pointer-types]
     ets_isr_attach(ETS_FRC_TIMER1_INUM, (func), (void *)(arg))
                                         ^
third-party/pwm/pwm.c:195:2: note: in expansion of macro 'ETS_FRC_TIMER1_INTR_ATTACH'
  ETS_FRC_TIMER1_INTR_ATTACH(pwm_intr_handler, NULL);
  ^
c:/Espressif/ESP8266_SDK/include/ets_sys.h:67:6: note: expected 'ets_isr_t {aka void (*)(void *)}' but argument is of type 'void (*)(void)'
 void ets_isr_attach(int i, ets_isr_t func, void *arg);
      ^
third-party/pwm/pwm.c: In function 'pwm_start':
c:/Espressif/ESP8266_SDK/include/ets_sys.h:99:5: error: implicit declaration of function 'ets_isr_mask' [-Werror=implicit-function-declaration]
     ets_isr_mask((1<<inum))
     ^
c:/Espressif/ESP8266_SDK/include/ets_sys.h:111:2: note: in expansion of macro 'ETS_INTR_DISABLE'
  ETS_INTR_DISABLE(ETS_FRC_TIMER1_INUM)
  ^
third-party/pwm/pwm.c:374:4: note: in expansion of macro 'ETS_FRC1_INTR_DISABLE'
    ETS_FRC1_INTR_DISABLE();
    ^
c:/Espressif/ESP8266_SDK/include/ets_sys.h:96:5: error: implicit declaration of function 'ets_isr_unmask' [-Werror=implicit-function-declaration]
     ets_isr_unmask((1<<inum))
     ^
c:/Espressif/ESP8266_SDK/include/ets_sys.h:108:2: note: in expansion of macro 'ETS_INTR_ENABLE'
  ETS_INTR_ENABLE(ETS_FRC_TIMER1_INUM)
  ^
third-party/pwm/pwm.c:391:3: note: in expansion of macro 'ETS_FRC1_INTR_ENABLE'
   ETS_FRC1_INTR_ENABLE();
   ^
cc1.exe: all warnings being treated as errors
make[1]: *** [compiler/lib/libpwm_open.a] Error 1
make[1]: Leaving directory `/C/DEV/ESP8266/Sming/Sming'
make: *** [/C/DEV/ESP8266/Sming/Sming/compiler/lib//libpwm_open.a] Error 2
```

I have changed one line in third-party/pwm/pwm.c
```
static void pwm_intr_handler()
->
static void pwm_intr_handler(void* ptr)
```
After that project was builded. But failed and rebooting every time:

```
3ffff770:  3ffe90ef 0000003c 0000001f 40100a11  
3ffff780:  4000050c 2de4a145 ffc2850c 00000014  
3ffff790:  3ffff8a1 3fffc100 40236628 400005e1  
3ffff7a0:  00000000 00000030 00000008 40236828  
3ffff7b0:  40236c54 00000000 00000028 00000000  
3ffff7c0:  3ffff8be 3ffff960 00000004 3ffff990  
3ffff7d0:  00000090 00000000 00000000 3ffe8370  
3ffff7e0:  0000001e 3ffff8a1 3ffff920 00000014  
3ffff7f0:  28ffeb21 00f00d02 c9f0c112 2c02cd21  
3ffff800:  01310902 00c082b0 0420c000 201132c0  
3ffff810:  c441c023 112280fe 424a030c 390020c0  
3ffff820:  fec24104 4a403551 0020c042 c0410459  
3ffff830:  c0224afe 02390020 3ffff844 3ffff854  
3ffff840:  3ffff850 0000001e 00f00d10 3ffef5ac  
3ffff850:  3ffff8be 00000080 3ffff970 3ffff960  
3ffff860:  00000004 3ffff960 0000000a 00000000  
3ffff870:  f00d10c1 180000ca 60000314 3ffe8380  
3ffff880:  00000000 a0000000 00000000 3ffff920  
3ffff890:  3ffff920 0001c200 0000007f 40236c2c  
3ffff8a0:  42454428 20294755 276e6f64 72702074  
3ffff8b0:  20746e69 6120656d 6c612074 75000a6c  
3ffff8c0:  6f206667 4155206e 0a305452 3ffff900  
3ffff8d0:  3ffff990 00000002 00000001 40001e4b  
3ffff8e0:  3ffff910 00000000 00000000 00000000  
3ffff8f0:  00000000 0000003a 00000001 2e2c2927  
3ffff900:  3ffef9a0 00000248 00000260 4021d344  
3ffff910:  4023855c 00000001 00000000 40006545  
3ffff920:  3ffff970 3ffff960 00000004 3ffff990  
3ffff930:  00000003 00000100 0001c200 3ffee134  
3ffff940:  3ffe9520 0001c200 3ffef508 40236c98  
3ffff950:  3ffff970 3ffff960 00000004 3ffee134  
3ffff960:  3ffe9520 402346ca 402346ca 0000000a  
3ffff970:  3ffff9ae 00000a6c 00000000 3ffee134  
3ffff980:  3ffe9520 0001c200 3ffef508 4023605b  
3ffff990:  42454428 20294755 276e6f64 72702074  
3ffff9a0:  20746e69 6120656d 6c612074 75000a6c  
3ffff9b0:  6f206667 4155206e 0a305452 ffffff00  
3ffff9c0:  402231c1 00000017 60000200 42d60000  
3ffff9d0:  00000000 3ffffab0 3ffffab0 3ffef877  
3ffff9e0:  3ffef928 0000ff00 00000000 40100e64  
3ffff9f0:  00000000 00000000 0000ff00 40100e64  
3ffffa00:  00000001 4021d36c 3ffef8f0 00000000  
3ffffa10:  3ffe90f0 3ffee134 00000000 00000028  
3ffffa20:  3ffe9189 3ffee134 00000028 4010642c  
3ffffa30:  40222996 00000000 3ffef8f0 40237502  
3ffffa40:  4021e53d 00000028 3ffef8f0 4023753a  
3ffffa50:  3ffef54c 3ffe90f0 3ffef8f0 4010655c  
3ffffa60:  3ffef938 00000000 00000000 4021d344  
3ffffa70:  00000000 00000001 00000002 00000003  
3ffffa80:  00000006 00000009 0000000c 0000000d  
3ffffa90:  0000000e 0000000f 00000014 0000001c  
3ffffaa0:  0000001d 40233a00 3ffe9318 402367d0  
3ffffab0:  40217a30 40233a00 3ffef830 402362e3  
3ffffac0:  4021821d 40218218 3ffef830 40218220  
3ffffad0:  3ffef830 00001000 4010077a 000003fd  
3ffffae0:  0000008a 00000020 00000022 00000000  
3ffffaf0:  3ffef8d8 00000000 00000003 450c1077  
3ffffb00:  ffffffff ffffffff ffff0001 0000000b  
3ffffb10:  746e6f44 6e6e6f43 00746365 00000000  
3ffffb20:  00000000 00000000 00000000 00000000  
3ffffb30:  00030503 69000003 726f6e67 656d4065  
3ffffb40:  00002b2b 00000000 00000000 00000000  
3ffffb50:  00000000 00000000 00000000 00000000  
3ffffb60:  00000000 00000000 00000000 00000000  
3ffffb70:  00000000 00000000 32618a25 ecf043a7  
3ffffb80:  4eb9c04d 042c64d1 c6b0e9ea 5bcd70c5  
3ffffb90:  6a6cbaec ce5b9d04 ffffffff ffffffff  
3ffffba0:  ffffffff ffffffff e38800ff 6c0d86ab  
3ffffbb0:  00000016 6e696d53 6e492067 6e726574  
3ffffbc0:  664f7465 6e696854 00007367 00000000  
3ffffbd0:  00000000 00000000 00000000 00000000  
3ffffbe0:  00000000 00000000 00000000 00000000  
3ffffbf0:  00000000 00000000 00000000 00000000  
3ffffc00:  00000000 00000000 00000000 00000000  
3ffffc10:  00000000 ffffff00 ffffffff ffffffff  
3ffffc20:  ffffffff ffffffff ffffffff ffffffff  
3ffffc30:  ffffffff 000007ff ffffff04 ffff0001  
3ffffc40:  0000000b 746e6f44 6e6e6f43 00746365  
3ffffc50:  00000000 00000000 00000000 00000000  
3ffffc60:  00000000 6f6e6769 6d406572 002b2b65  
3ffffc70:  00000000 00000000 00000000 00000000  
3ffffc80:  00000000 00000000 00000000 00000000  
3ffffc90:  00000000 00000000 00000000 00000000  
3ffffca0:  00000000 ffffffff ffffffff ffffffff  
3ffffcb0:  ffffffff ffffffff ffffffff ffffffff  
3ffffcc0:  ffffffff ffffffff ffffffff ffffffff  
3ffffcd0:  ffffffff ffffffff ffffffff ffffffff  
3ffffce0:  ffffffff ffffffff ffffffff ffffffff  
3ffffcf0:  ffffffff ffffffff ffffffff ffffffff  
3ffffd00:  ffffffff ffffffff ffffffff ffffffff  
3ffffd10:  ffffffff ffffffff ffffffff ffffffff  
3ffffd20:  ffffffff ffffffff ffffffff ffffffff  
3ffffd30:  ffffffff ffffffff ffffffff ffffffff  
3ffffd40:  ffffffff ffffffff ffffffff ffffffff  
3ffffd50:  ffffffff ffffffff ffffffff ffffffff  
3ffffd60:  ffffffff ffffffff ffffffff ffffffff  
3ffffd70:  ffffffff ffffffff ffffffff ffffffff  
3ffffd80:  ffffffff ffffffff ffffffff ffffffff  
3ffffd90:  ffffffff ffffffff ffffffff ffffffff  
3ffffda0:  ffffffff ffffffff ffffffff ffffffff  
3ffffdb0:  ffffffff ffffffff ffffffff ffffffff  
3ffffdc0:  ffffffff ffffffff ffffffff ffffffff  
3ffffdd0:  ffffffff ffffffff ffffffff ffffffff  
3ffffde0:  ffffffff ffffffff ffffffff ffffffff  
3ffffdf0:  ffffffff ffffffff ffffffff ffffffff  
3ffffe00:  ffffffff ffffffff ffffffff ffffffff  
3ffffe10:  ffffffff ffffffff ffffffff ffffffff  
3ffffe20:  ffffffff ffffffff ffffffff ffffffff  
3ffffe30:  ffffffff 04080000 001c110c 01010000  
3ffffe40:  ffffff01 00000003 ffffff00 ffffffff  
3ffffe50:  ffffffff ffffffff ffffffff ffffffff  
3ffffe60:  ffffffff ffffffff ffffffff ffff00c8  
3ffffe70:  00000000 00000000 ffffffff ffffffff  
3ffffe80:  ffffffff ffffffff ffffffff ffffffff  
3ffffe90:  ffffffff ffffffff ffffffff ffffffff  
3ffffea0:  ffffffff ffffffff ffffffff ffffffff  
3ffffeb0:  ffffffff ffffffff ffffffff ffffffff  
3ffffec0:  ffffffff ffffffff ffffffff ffffffff  
3ffffed0:  ffffffff ffffffff ffffffff ffffffff  
3ffffee0:  ffffffff 32618a25 ecf043a7 4eb9c04d  
3ffffef0:  042c64d1 c6b0e9ea 5bcd70c5 6a6cbaec  
3fffff00:  ce5b9d04 ffffffff ffffffff ffffffff  
3fffff10:  ffffffff ffffffff ffffffff ffffffff  
3fffff20:  ffffffff ffffffff ffffffff ffffffff  
3fffff30:  ffffffff ffffffff ffffffff ffffffff  
3fffff40:  ffffffff ffffffff ffffffff ffffffff  
3fffff50:  ffffffff ffffffff ffffffff ffffffff  
3fffff60:  ffffffff ffffffff ffffffff ffffffff  
3fffff70:  ffffffff ffffffff ffffffff ffffffff  
3fffff80:  ffffffff ffffffff ffffffff ffffffff  
3fffff90:  fe18ffff 51cff134 3ffef740 ab000000  
3fffffa0:  400203e9 40004928 ffffff00 55aa55aa  
```

**Another point**
I think will be right to include Espressif SDK as other 3-rd part module to Sming (axtls-8266, rboot, spiffs...)
It will decerease chances of compilation issues, and also may be can make Sming more stable because of tested distributive.

slaff commented:
@anakod Try the following:
```
cd %SMING_HOME%
make dist-clean
make
````

anakod commented:
Unfortunately, this didn't help, same result :(

anakod commented:
**@slaff, I'm very sorry for distortion because of insufficiently deep testing :(** Usual I'm always start any tests from Basic_Serial project, because in my opinion, it should be simplest example and should work anyway. So when it didn't work - I think what nothing work :). As result of unsuccessful test I have decided to remove and re-install everything related to ESP on my laptop from zero.
It didn't help me again, and I still have same issue, so no results. After that, I decide to try Basic_Blink example ... and it works!
So probably last above issue related only to Serial example (after all my experiments it still didn't work, we need to look it separetely, I think).

Next, I have decided to test more complex HttpServer_Bootstrap example to check Sming stability. There was a few new issues:

1. I can't compile it because of "undefined reference to `__assert_func'" error. As a workaround I have created custom assert.h file, which will overwrite assert() function for all external libraries. It helps because these libraries don't use actual Sming includes directly (and user_config.h at first).
I have pushed related commit to "slaff-update/sdk-2.1", but you are free to revert it, if you think what it's wrong approach (and it may be!). From my point of view it can be a nice workaround hack, because we "redefine" standard C functionality for our SoC architecture without any code chages in external libraries.

2. Every ~5-10 seconds module lost connection and reconect to WiFi network:
```
connected with SSID, channel 1
dhcp client start...
connect to ssid SSID, channel 1

ip:192.168.7.61,mask:255.255.255.0,gw:192.168.7.1
ip:192.168.7.61,mask:255.255.255.0,gw:192.168.7.1


'/' registered
'/hello' registered
'*' registered

=== WEB SERVER STARTED ===
192.168.7.61
==============================

pm open,type:2 0
bcn_timout,ap_probe_send_start
ap_probe_send over, rest wifi status to disassoc
state: 5 -> 0 (1)
rm 0
pm close 7
disconnect from ssid SSID, reason 200

scandone
no SSID found, reconnect after 1s
disconnect from ssid SSID, reason 201

reconnect
scandone
state: 0 -> 2 (b0)
state: 2 -> 3 (0)
state: 3 -> 5 (10)
add 0
aid 8
cnt 

connected with SSID, channel 1
dhcp client start...
connect to ssid SSID, channel 1

ip:192.168.7.61,mask:255.255.255.0,gw:192.168.7.1
ip:192.168.7.61,mask:255.255.255.0,gw:192.168.7.1


'/' registered
'/hello' registered
'*' registered

=== WEB SERVER STARTED ===
192.168.7.61
==============================
```

I have tryed to google it, probably this is related issue, but still no answer:
https://esp32.com/viewtopic.php?t=61

After that, I have tryed to change WiFi network Beacon Interval in my router configuration page from default value "100" to "5". And this issue was resolved. (this isn't good, we need to know how to configure beacon value on ESP to keep router settings without any changes, I think)
In my opinion, it's related to SDK update, but router firmware was updated some time before also, so I'm only 80% sure about that. I will try to check on another router, I will back to you after that.

Next, looks like we have some strange issue with filesystem initialization. On first start with clean flash (after `make flashinit`), ESP always "formating" filesystem. After that, if I'm trying to reboot module ESP can't start anymore (print many random charecters in the console)
If, after that, I'm executing "esptool.exe -p COM6 -b 256000 write_flash -ff 40m -fm dio -fs 32m 0x3fc000 esp_init_data_default.bin" - module will successfuly start again after next reboot.
So I think filesystem position and\or size detection code have some issue, **or** something wrong with my configuration (but I think, I have not such issue before update, not sure how to check now)

Just in case, my current Environment configuration is:
https://gyazo.com/984195ceee43a9ed86d7d767e9d1490a

Next, I decide to test one of the most complex examples - HttpServer_ConfigNetwork. Unfortunately it didn't work as it should. I was able to conrnect ESP Access Point, but haven't ability to configure Station mode connection for existing network.

Next I have tryed FtpServer example. There was same issue as in HttpServer_Bootstrap - I need to manually flash esp_init_data_default.bin to be able to restart module after filesystem formatting. Good news, what after that I was able to download\upload even big files and everything works very fine (only issue what download speed was limited to 4Kb\sec by some reason, upload speed - wasn't).
I'm sure there was few related bugs in earlier versions, so FTP works much better in comparsion with early Sming builds.

slaff commented:
@anakod Thanks for the testing. I modified a bit the code and the travis tests and as far as I can see only the example with enabled GDB is the one that is not compiling( the issue is [reported](https://github.com/espressif/esp-gdbstub/issues/17)). Our travis system uses the compiler produced from esp-open-sdk. 

Can you also test with the `esp-open-sdk` compiler?

slaff commented on Sming/system/uart.cpp:
```diff
@@ -492,16 +492,16 @@ void uart_set_debug(int uart_nr)
     case UART0:
         system_set_os_print(1);
-        ets_install_putc1((void *) &uart0_write_char);
+        ets_install_putc1((void (*)(char)) &uart0_write_char);
```
That should be simplified to something like:  `ets_install_putc1(uart0_write_char);`.

slaff commented on Sming/system/uart.cpp:
```diff
@@ -140,7 +140,7 @@ void uart_start_isr(uart_t* uart)
     USIE(uart->uart_nr) = (1 << UIFF) | (1 << UIFR) | (1 << UITO);
 //    ETS_UART_INTR_ATTACH(uart_isr,  (void *)uart);
-    ETS_UART_INTR_ATTACH((void *)uart_isr,  (void *)uart);
+    ETS_UART_INTR_ATTACH((ets_isr_t)uart_isr,  (void *)uart);
```
The casting `(ets_isr_t)uart_isr` is not really needed.

slaff commented on Sming/system/uart.cpp:
```diff
@@ -492,16 +492,16 @@ void uart_set_debug(int uart_nr)
     case UART1:
         system_set_os_print(1);
-        ets_install_putc1((void *) &uart1_write_char);
+        ets_install_putc1((void (*)(char)) &uart1_write_char);
```
That should be simplified to something like: ets_install_putc1(uart1_write_char);

slaff commented:
> I think will be right to include Espressif SDK as other 3-rd part module to Sming (axtls-8266, rboot, spiffs...) ...

@anakod I added this suggestion as part of the PR. For the moment the code and samples can compile just fine. With the exception of ESP-Gdbstub (reported [here](https://github.com/espressif/esp-gdbstub/issues/17) ). Which means that direct debugging on the device is broken until the right function definition is provided.

slaff commented:
Most of the code seems to compile now but there is one disturbing issue. Somehow 4K of RAM got consumed by the new SDK

Comparison with `HttpServer_ConfigNetwork'`

#SDK 2.1 Memory / Section info:
```
------------------------------------------------------------------------------
   Section|                   Description| Start (hex)|   End (hex)|Used space
------------------------------------------------------------------------------
      data|        Initialized Data (RAM)|    3FFE8000|    3FFE8384|     900
    rodata|           ReadOnly Data (RAM)|    3FFE8390|    3FFEB284|   12020
       bss|      Uninitialized Data (RAM)|    3FFEB288|    3FFF2088|   28160
      text|            Cached Code (IRAM)|    40100000|    40106BE4|   27620
irom0_text|           Uncached Code (SPI)|    4020A000|    4025086C|  288876
Total Used RAM : 41080
Free RAM : 40840
Free IRam : 5166
------------------------------------------------------------------------------
```
#SDK 1.5 Memory / Section info:

```
------------------------------------------------------------------------------
   Section|                   Description| Start (hex)|   End (hex)|Used space
------------|-------------------------------|----------------|--------|---------
      data|        Initialized Data (RAM)|    3FFE8000|    3FFE8384|     900
    rodata|           ReadOnly Data (RAM)|    3FFE8388|    3FFEB28C|   12036
       bss|      Uninitialized Data (RAM)|    3FFEB290|    3FFF1FD0|   27968
      text|            Cached Code (IRAM)|    40100000|    40105AA3|   23203
irom0_text|           Uncached Code (SPI)|    40209000|    40249B36|  265014
------------------------------------------------------------------------------
Total Used RAM : 40904
Free RAM : 41016
Free IRam : 9583
```

Any ideas what is wrong here?

slaff commented:
@anakod Some progress information:

 - Basic_Serial started to work as expected after I reverted the latest change to `m_printf.cpp`, introduced in ed94fb77d7e8a948e7fc6837844b48d1a94f933c, the sample worked as expected. There are other issues most probably related to this commit (#1274) so I will revert it for the `develop` branch.

- HttpServer_Bootstrap -
> 1. disconnect from ssid SSID, reason 201
> 2.  reboot module ESP can't start anymore

I have experienced the same issues as you. Which means that for now we cannot use the new SDK until we figure out:
- [ ]   the reason for the constant disconnects. 
- [ ]  why we need to reflash the `esp_init_data_default.bin` 

Concerning the `HttpServer_ConfigNetwork` sample: 
 
- HttpServer_ConfigNetwork - Initially I had the same results as you. Then I flashed the  `esp_init_data_default.bin` file from SDK 1.5. And the example started to work without any glitches. 

@anakod Any thoughts on this?

kwis2 commented:
with SDK Version 2.1.0(116b762) and latest sming I got:

> 
> CC system/crash_handler.c
> In file included from include/user_config.h:33:0,
>                  from system/crash_handler.c:24:
> system/include/esp_systemapi.h:48:13: error: conflicting types for 'ets_timer_ar
> m_new'
>  extern void ets_timer_arm_new(ETSTimer *ptimer, uint32_t milliseconds, bool rep
> eat_flag, int isMstimer);
>              ^
> In file included from system/include/esp_systemapi.h:7:0,
>                  from include/user_config.h:33,
>                  from system/crash_handler.c:24:
> c:/Espressif/ESP8266_SDK/include/osapi.h:65:6: note: previous declaration of 'et
> s_timer_arm_new' was here
>  void ets_timer_arm_new(os_timer_t *ptimer, uint32_t time, bool repeat_flag, boo
> l ms_flag);
>       ^
> In file included from include/user_config.h:33:0,
>                  from system/crash_handler.c:24:
> system/include/esp_systemapi.h:57:13: error: conflicting types for 'ets_delay_us
> '
>  extern void ets_delay_us(uint32_t us);
>              ^
> In file included from system/include/esp_systemapi.h:7:0,
>                  from include/user_config.h:33,
>                  from system/crash_handler.c:24:
> c:/Espressif/ESP8266_SDK/include/osapi.h:33:6: note: previous declaration of 'et
> s_delay_us' was here
>  void ets_delay_us(uint16_t us);
>       ^
> make[1]: *** [out/build/system/crash_handler.o] Error 1
> make[1]: Leaving directory `/C/tools/sming/Sming'
> make: *** [reload] Error 2
> 
> C:\tools\sming\Sming>

question which SDK version should I install to work with sming?
